ruby_event_store/lib/ruby_event_store/mappers/transformation/preserve_types.rb

```diff
@@ -0,0 +1,143 @@
+# frozen_string_literal: true
+
+module RubyEventStore
+  module Mappers
+    module Transformation
+      SerializationError = Class.new(ArgumentError)
+
+      class PreserveTypes
+        def initialize(type_resolver: ->(type) { type.to_s } )
+          @registered_type_serializers = {}
+          @type_resolver = type_resolver
+        end
+
+        def register(type, serializer:, deserializer:)
+          @registered_type_serializers[@type_resolver.(type)] = {
+            serializer: serializer,
+            deserializer: deserializer
+          }
+          self
+        end
+
+        def dump(record)
+          data = transform(record.data)
+          metadata = transform(record.metadata)
+          if (metadata.respond_to?(:[]))
```

Comment:
can we make an assumption on Metadata being always in an Event?

Reply:
With `RubyEventStore::Event` this assumption is always true because of its initializer implementation

```ruby
    def initialize(event_id: SecureRandom.uuid, metadata: nil, data: {})
      @event_id = event_id.to_s
      @metadata = Metadata.new(metadata.to_h)
      @data     = data
    end
 ```
 
 If we prepare event with nil metadata (or use another event class that allows that) we will get:
 
 ```ruby
 > e = RubyEventStore::Event.new
=> #<RubyEventStore::Event:0x00007fc630389408 @event_id="2755b447-78d1-4dad-aec3-29792a9b6567", @metad...

> e.metadata
=> #<RubyEventStore::Metadata:0x00007fc630388eb8 @h={}>

> e.instance_variable_set("@metadata", nil)
=> nil
> e.metadata
=> nil

> c = RubyEventStore::Client.new(repository: RubyEventStore::InMemoryRepository.new)
=> #<RubyEventStore::Client:0xb4>

> c.publish(e)
Traceback (most recent call last):
        9: from /Users/mpraglowski/.rbenv/versions/2.7.2/bin/irb:23:in `<main>'
        8: from /Users/mpraglowski/.rbenv/versions/2.7.2/bin/irb:23:in `load'
        7: from /Users/mpraglowski/.rbenv/versions/2.7.2/lib/ruby/gems/2.7.0/gems/irb-1.2.6/exe/irb:11:in `<top (required)>'
        6: from (irb):13
        5: from /Users/mpraglowski/arkency/rails_event_store/ruby_event_store/lib/ruby_event_store/client.rb:32:in `publish'
        4: from /Users/mpraglowski/arkency/rails_event_store/ruby_event_store/lib/ruby_event_store/client.rb:341:in `enrich_events_metadata'
        3: from /Users/mpraglowski/arkency/rails_event_store/ruby_event_store/lib/ruby_event_store/client.rb:341:in `each'
        2: from /Users/mpraglowski/arkency/rails_event_store/ruby_event_store/lib/ruby_event_store/client.rb:341:in `block in enrich_events_metadata'
        1: from /Users/mpraglowski/arkency/rails_event_store/ruby_event_store/lib/ruby_event_store/client.rb:347:in `enrich_event_metadata'
NoMethodError (undefined method `[]' for nil:NilClass)
```

So here in transformations I could assume metadata are present. The question is if I should make such assumptions for both data & metadata ;)

Reply:
BTW this works because:

```ruby
> nil.respond_to?(:[])
=> false
```